### ISSUE TITLE
nushell should be overriding global keybinds like "up arrow" with local keybinds

### ISSUE THREAD

**Issue by freijon** [nushell/nushell#10414]
### Describe the bug

See also: https://github.com/atuinsh/atuin/issues/1025

I'm using `atuin` to manage my command history. Their init script binds up-arrow to show the command history. Unfortunately, now `atuin` is triggered every time when I press up-arrow, even when I don't mean to.

### How to reproduce

1. Run nushell with ootb command history
2. type `math <TAB>` to trigger completion
3. Cycle through the options with arrow-up / arrow-down --> It works 🎉
4. Now install `atuin` and source [atuin.nu](https://github.com/atuinsh/atuin/blob/main/atuin/src/shell/atuin.nu)
5. type `math <TAB>` to trigger completion
6. Cycle through the options with arrow-up / arrow-down --> When pressing up-arrow `atuin` is triggered

### Expected behavior

The local keybind should only be triggered when not in completion mode

### Configuration

| key                | value                                         |
| ------------------ | --------------------------------------------- |
| version            | 0.84.0                                        |
| branch             |                                               |
| commit_hash        |                                               |
| build_os           | linux-x86_64                                  |
| build_target       | x86_64-unknown-linux-gnu                      |
| rust_version       | rustc 1.71.1 (eb26296b5 2023-08-03) (gentoo)  |
| cargo_version      | cargo 1.71.1                                  |
| build_time         | 2023-08-28 08:14:33 +02:00                    |
| build_rust_channel | release                                       |
| allocator          | standard                                      |
| features           | dataframe, default, sqlite, trash, which, zip |
| installed_plugins  | from-binary                                   |

**Comment by fdncred** [comment]
When I noticed that up arrow was bound to atuin, I immediately went to my init.nu for atuin and commented out that keybinding. It's annoying to me and way to slow when I just want to get to the last few commands.

Having said this, I tested atuin again on Windows (without commenting out the up arrow keybinding), and don't notice the behavior you're describing. Once I'm in the atuin UI, I can navigate around with up/down arrows without issue.

**Comment by freijon** [comment]
Yes, when inside the atuin UI up/down works as expected. I'm talking about nushell completions. Start typing "math", press TAB and naviate with up/down.

**Comment by fdncred** [comment]
OH! I see what you mean. That's some bizarre behavior. Sorry that I misunderstood. I guess I never noticed because I commented out that keybinding and use tab/shift-tab to navigate completions.

I'm not sure how to do what you're saying. It kind of pre-supposes that nushell has keybindings for nushell menus and a set of other keybindings for other things. I don't believe this is possible with nushell right now. We have one set of keybindings. I'd be willing to look at a PR for menu-keybindings + non-menu-keybindings though.

**Comment by arcuru** [comment]
The way other shells (fish) do this is by allowing a keybinding to query the state using an interface like [commandline](https://www.nushell.sh/commands/docs/commandline.html). The fish [commandline](https://fishshell.com/docs/current/cmds/commandline.html) utility does this by providing access to the `mode` state (`--search-mode`/`--paging-mode`) and access to the line number of the command (so we can know which line of text the cursor is currently on).

See here for how we write it for fish: https://github.com/atuinsh/atuin/blob/b74ed756e2542d330be81ee841407fe263453d42/atuin/src/shell/atuin.fish#L33-L49

Nushell could support this by adding similar functionality to [commandline](https://www.nushell.sh/commands/docs/commandline.html), and then we could use that in the keybinding to determine whether to take any action. Atuin only needs to know if the cursor is on the topmost line and that is probably enough.

**Comment by fdncred** [comment]
@arcuru We'd probably accept a PR for this type of change to `commandline`.

**Comment by remmycat** [comment]
Hey, I was wondering if this issue is fixed by keybindings having the [`until` functionality](https://www.nushell.sh/book/line_editor.html#edit-type)?

That is how the issue was solved on atuin's side in https://github.com/atuinsh/atuin/pull/1770 , basically changing the up-arrow keybinding's `event` from 

this version (shortened for brevity)

```nushell
{
  name: atuin
  modifier: none
  keycode: up
  mode: [emacs, vi_normal, vi_insert]
  event: {
    send: executehostcommand
    cmd: (_atuin_search_cmd '--shell-up-key-binding')
  }
}
```

to this version, that first tries to interact with a menu when one is open:

```nushell
{
  name: atuin
  modifier: none
  keycode: up
  mode: [emacs, vi_normal, vi_insert]
  event: {
    until: [
      {send: menuup}
      {
        send: executehostcommand
        cmd: (_atuin_search_cmd '--shell-up-key-binding')
      }
    ]
  }
}
```

**Comment by 132ikl** [comment]
seems like this is equivalent to https://github.com/atuinsh/atuin/issues/1025. I'll close this, if you are still having this issue then let me know and I can reopen it.
